### DISCUSSION TITLE
Get number of dims at EP GetCapability() time

### DISCUSSION THREAD

**Discussion by eralmual** [microsoft/onnxruntime#16258] [Other Q&A]
Hi all! 

We are trying to implement new features into our EP but for them we need to know either the tensor dimensions or number of dimensions when the _GetCapability_ method gets called. 
We understand that knowing the specific shapes of a dynamic model at that time is pretty much impossible, but we where wondering if there was a way to at least access the number of dimensions as this information can be found in the ONNX model.

For example, we've seen that dynamic models have input shapes like [batch_size, width, height] in that case we would like to be able to know that num_dims=3.

Currently is there any way to do this?

**Comment by edgchen1** [comment]
Yes, given a `Node`, you can check the available shape information in its input `NodeArg`s (from `Node::InputDefs`). Here's an example:

https://github.com/microsoft/onnxruntime/blob/d1e8d4a2613f25823d997a38d51b138ca9adcedb/onnxruntime/core/providers/coreml/builders/impl/depthtospace_op_builder.cc#L63-L73

https://github.com/microsoft/onnxruntime/blob/d1e8d4a2613f25823d997a38d51b138ca9adcedb/onnxruntime/core/providers/coreml/builders/helper.cc#L22-L35

**Reply by eralmual** [reply]
Is the result of _node_arg.Shape()_ always nullptr when we have dynamic shapes? Is there no way to know the number of dims of a dynamic tensor? It's in this case we are particularly interested, because we can extract the shapes from constant tensors but we also need the number of dims in the dynamic ones to optimize the op
